tools/datagen/src/bin/verify-zero-copy.rs

```diff
@@ -0,0 +1,257 @@
+            Arg::with_name("KEYS")
+                .short("k")
+                .long("keys")
+                .multiple(true)
+                .takes_value(true)
+                .help(
+                    "Include this resource key in the output. Accepts multiple arguments. \
+                    Also see --test-keys.",
+                ),
+        )
+        .group(
+            ArgGroup::with_name("KEY_MODE")
+                .arg("KEYS")
+                .arg("TEST_KEYS")
+                .required(true),
+        )
+        .arg(
+            Arg::with_name("TEST_KEYS")
+                .long("test-keys")
+                .help("Include all keys supported by testdata."),
+        )
+        .get_matches();
+
+    if matches.is_present("VERBOSE") {
+        SimpleLogger::new()
+            .with_level(log::LevelFilter::Trace)
+            .init()
+            .unwrap()
+    } else {
+        SimpleLogger::new()
+            .env()
+            .with_level(log::LevelFilter::Info)
+            .init()
+            .unwrap()
+    }
+
+    let all_keys = get_all_keys();
+    let selected_keys = if matches.is_present("TEST_KEYS") {
+        // We go with all keys now and later ignore key errors.
+        all_keys
+    } else {
+        let mut keys = HashSet::new();
+
+        if let Some(paths) = matches.values_of("KEYS") {
+            keys.extend(paths.map(Cow::Borrowed));
+        }
+
+        let filtered: Vec<_> = all_keys
+            .into_iter()
+            .filter(|k| keys.contains(k.get_path()))
+            .collect();
+
+        if filtered.is_empty() {
+            eyre::bail!("No keys selected (or keys passed in do not exist), pass in --keys KEYNAME or --test-keys");
+        }
+
+        filtered
+    };
+
+    let cldr_json_root = if let Some(cldr) = matches.value_of("CLDR_ROOT") {
+        PathBuf::from(cldr)
+    } else if matches.is_present("INPUT_FROM_TESTDATA") {
+        icu_testdata::paths::cldr_json_root()
+    } else {
+        eyre::bail!("Value for --cldr--root must be specified (or --input-from-testdata)",)
+    };
+    let uprops_root = if let Some(uprops) = matches.value_of("UPROPS_PATH") {
+        PathBuf::from(uprops)
+    } else if matches.is_present("INPUT_FROM_TESTDATA") {
+        icu_testdata::paths::uprops_toml_root()
+    } else {
+        eyre::bail!("Value for --uprops-path must be specified (or --input-from-testdata)",)
+    };
+    let cldr_paths = Box::new(CldrPathsAllInOne {
+        cldr_json_root,
+        locale_subset: matches
+            .value_of("CLDR_LOCALE_SUBSET")
+            .unwrap_or("full")
+            .to_string(),
+    });
+
+    let converter: Box<dyn OmnibusDatagenProvider<SerializeMarker> + Sync> =
+        Box::new(ForkByKeyProvider(
+            icu_provider_cldr::create_exportable_provider(
+                cldr_paths.as_ref(),
+                uprops_root.clone(),
+            )?,
+            icu_provider_uprops::create_exportable_provider(&uprops_root)?,
+        ));
```

Comment:
I don't like the `HeapStatsMarker`/`convert` design. The raw CLDR/uprops data should not be required by this test, and yet it is. Also it adds the benchmarking code and dependency in the `icu_provider` crate, even if it's behind feature, I think it should really all be in this crate.

Reply:
About the CLDR/uprops data: it is not actually required for anything; it is only here because the impl for DataConverter is on CldrJsonDataprovider and the various PropertiesProviders, and the only way to get that impl is to construct them. Better would be if we could get a "static" instance, since all we need is the type info.

About the location of the benchmarking code: the other way to do it would be to have a callback function. It gets rid of the optional dehat dependency in icu_provider, which is nice. But I like the DataConverter design too (Manish and I designed it on a whiteboard). If you think you can do better with a callback function, you could make an attempt after this PR is landed (after finishing the OKR goals).

Reply:
It can't be since the registry code needs to be everywhere. I did have a registry design that was more generic but it had some problems and we opted for this which is more modular (and allows for the crab bake registry, for example, to not handle all types)

You're right that the raw data isn't super needed here. We can potentially add modes to these providers that choose to not load data but I'm unsure if we should.

Overall the `convert()` design is needed for a bunch of other things too, that's why we picked it.